proxy/http3/Http3Types.h

```diff
@@ -95,29 +95,37 @@ enum class Http3ErrorCode : uint16_t {
   QPACK_DECODER_STREAM_ERROR = 0x202,
 };
 
+enum class Http3ErrorType { UNDEFINED = 0, CONNECTION_ERROR, STREAM_ERROR };
+
 class Http3Error
 {
 public:
   virtual ~Http3Error() {}
-  uint16_t code();
 
   Http3ErrorClass cls = Http3ErrorClass::NONE;
-  union {
-    Http3ErrorCode app_error_code;
-  };
-  const char *msg = nullptr;
+  Http3ErrorCode code = Http3ErrorCode::H3_NO_ERROR;
+  const char *msg     = nullptr;
+  Http3ErrorType type = Http3ErrorType::UNDEFINED;
+
+  uint16_t
+  get_code()
+  {
+    return static_cast<uint16_t>(code);
+  }
 
 protected:
-  Http3Error() : app_error_code(Http3ErrorCode::H3_NO_ERROR){};
-  Http3Error(const Http3ErrorCode error_code, const char *error_msg = nullptr)
-    : cls(Http3ErrorClass::APPLICATION), app_error_code(error_code), msg(error_msg){};
+  Http3Error() : code(Http3ErrorCode::H3_NO_ERROR){};
+  Http3Error(const Http3ErrorCode error_code, const char *error_msg = nullptr,
+             const Http3ErrorType error_type = Http3ErrorType::UNDEFINED)
+    : cls(Http3ErrorClass::APPLICATION), code(error_code), msg(error_msg), type(error_type){};
 };
 
 class Http3ConnectionError : public Http3Error
 {
 public:
   Http3ConnectionError() : Http3Error() {}
-  Http3ConnectionError(const Http3ErrorCode error_code, const char *error_msg = nullptr) : Http3Error(error_code, error_msg){};
+  Http3ConnectionError(const Http3ErrorCode error_code, const char *error_msg = nullptr)
+    : Http3Error(error_code, error_msg, Http3ErrorType::CONNECTION_ERROR){};
```

Comment:
If we are going to have error type, we no longer need this sub class (and the one for stream as well).